tsl/src/remote/dist_txn.c

```diff
@@ -177,36 +221,56 @@ dist_txn_xact_callback_1pc_pre_commit()
 		async_request_set_add(ars, remote_txn_async_send_commit(remote_txn));
 	}
 
-	testing_callback_call("waiting-commit");
+	eventcallback(DTXN_EVENT_WAIT_COMMIT);
 
 	/* async collect all the replies */
 	async_request_set_wait_all_ok_commands(ars);
 
 	dist_txn_deallocate_prepared_stmts_if_needed();
 }
 
-/* On abort on the frontend send aborts to all of the remote endpoints */
+/*
+ * Abort on the access node.
+ *
+ * The access node needs to send aborts to all of the remote endpoints. This
+ * code should not throw errors itself, since we are already in abort due to a
+ * previous error. Instead, we try to emit errors as warnings. For safety, we
+ * should probaby try-catch and swallow any potential lower-layer errors given
+ * that we're doing remote calls over the network. But the semantics for
+ * capturing and proceeding after such recursive errors are unclear.
+ */
 static void
 dist_txn_xact_callback_abort()
 {
 	RemoteTxn *remote_txn;
 
-	testing_callback_call("pre-abort");
+	eventcallback(DTXN_EVENT_PRE_ABORT);
+
 	remote_txn_store_foreach(state.store, remote_txn)
 	{
-		if (!remote_txn_is_ongoing(remote_txn) || !remote_txn_abort(remote_txn))
+		/*
+		 * Make sure we do not throw errors here because that will call the abort
+		 * handler again, causing an infinite loop until stack memory runs out.
+		 */
+		bool txn_is_ongoing = remote_txn_is_ongoing(remote_txn);
+
+		if (!txn_is_ongoing || !remote_txn_abort(remote_txn))
 		{
-			if (remote_txn_is_ongoing(remote_txn))
+			if (txn_is_ongoing)
 				elog(WARNING, "failure aborting remote transaction during local abort");
+			else
+				elog(WARNING, "no ongoing remote transactions during local abort");
```

Comment:
Is it a `WARNING` or `NOTICE`?

Reply:
Good question. Maybe we shouldn't warn at all here, like the code was previously. But I don't understand why we are only removing the transaction if it is ongoing @cevian?

It seems to me that if the local transaction failed, all remote transaction should fail too and be removed. So, the only question here is whether the remote transaction is ongoing and needs an abort or can just be removed because it was never started on the remote end.

Reply:
I guess a more intuitive approach here is to iterate the transactions and cancel the ongoing ones. 

Then clear the store in one swoop at the end instead of in-loop removal (and only in some cases).

Reply:
The connection cache was refactored in a previous PR, so resolving this convo.